File: dp/c/cache_aside.c
// Cache-Aside Pattern in C

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX_CACHE 16

typedef struct { char key[32]; char value[64]; bool valid; } CacheEntry;

typedef struct { CacheEntry entries[MAX_CACHE]; int count; } Cache;

void cache_init(Cache* c) { c->count = 0; }

const char* cache_get(Cache* c, const char* key) {
    for (int i = 0; i < c->count; i++)
        if (c->entries[i].valid && strcmp(c->entries[i].key, key) == 0) {
            printf("[Cache] HIT: %s\n", key);
            return c->entries[i].value;
        }
    printf("[Cache] MISS: %s\n", key);
    return NULL;
}

void cache_put(Cache* c, const char* key, const char* value) {
    for (int i = 0; i < c->count; i++)
        if (strcmp(c->entries[i].key, key) == 0) {
            snprintf(c->entries[i].value, 64, "%s", value);
            c->entries[i].valid = true;
            return;
        }
    snprintf(c->entries[c->count].key, 32, "%s", key);
    snprintf(c->entries[c->count].value, 64, "%s", value);
    c->entries[c->count].valid = true;
    c->count++;
    printf("[Cache] PUT: %s=%s\n", key, value);
}

void cache_invalidate(Cache* c, const char* key) {
    for (int i = 0; i < c->count; i++)
        if (strcmp(c->entries[i].key, key) == 0)
            c->entries[i].valid = false;
    printf("[Cache] INVALIDATE: %s\n", key);
}

// Simulated DB
const char* db_query(const char* key) {
    printf("[DB] Query: %s\n", key);
    if (strcmp(key, "user:1") == 0) return "Alice";
    if (strcmp(key, "user:2") == 0) return "Bob";
    return NULL;
}

const char* get_user(Cache* c, const char* key) {
    const char* val = cache_get(c, key);
    if (val) return val;
    val = db_query(key);
    if (val) cache_put(c, key, val);
    return val;
}

int main(void) {
    Cache cache;
    cache_init(&cache);

    printf("=== First access ===\n");
    printf("Result: %s\n\n", get_user(&cache, "user:1"));

    printf("=== Second access (cached) ===\n");
    printf("Result: %s\n\n", get_user(&cache, "user:1"));

    printf("=== After invalidation ===\n");
    cache_invalidate(&cache, "user:1");
    printf("Result: %s\n", get_user(&cache, "user:1"));
    return 0;
}

File: dp/rust/src/cache_aside.rs
// Cache-Aside Pattern in Rust
// Application manages cache explicitly: read-through with lazy population

use std::collections::HashMap;
use std::time::{Duration, Instant};

struct CacheEntry<T> {
    value: T,
    expires_at: Instant,
}

struct Cache<T> {
    entries: HashMap<String, CacheEntry<T>>,
    ttl: Duration,
}

impl<T: Clone + std::fmt::Debug> Cache<T> {
    fn new(ttl: Duration) -> Self {
        Cache {
            entries: HashMap::new(),
            ttl,
        }
    }

    fn get(&self, key: &str) -> Option<&T> {
        self.entries.get(key).and_then(|entry| {
            if Instant::now() < entry.expires_at {
                println!("[Cache] HIT: {}", key);
                Some(&entry.value)
            } else {
                println!("[Cache] EXPIRED: {}", key);
                None
            }
        })
    }

    fn put(&mut self, key: String, value: T) {
        println!("[Cache] PUT: {} = {:?}", key, value);
        self.entries.insert(key, CacheEntry {
            value,
            expires_at: Instant::now() + self.ttl,
        });
    }

    fn invalidate(&mut self, key: &str) {
        println!("[Cache] INVALIDATE: {}", key);
        self.entries.remove(key);
    }
}

// Simulated database
struct Database {
    data: HashMap<String, String>,
}

impl Database {
    fn new() -> Self {
        let mut data = HashMap::new();
        data.insert("user:1".into(), "Alice".into());
        data.insert("user:2".into(), "Bob".into());
        data.insert("user:3".into(), "Charlie".into());
        Database { data }
    }

    fn query(&self, key: &str) -> Option<String> {
        println!("[DB] Query: {}", key);
        self.data.get(key).cloned()
    }
}

// Cache-aside service
fn get_user(cache: &mut Cache<String>, db: &Database, key: &str) -> Option<String> {
    // 1. Check cache
    if let Some(val) = cache.get(key) {
        return Some(val.clone());
    }

    // 2. Cache miss — load from DB
    println!("[Cache] MISS: {}", key);
    if let Some(val) = db.query(key) {
        // 3. Populate cache
        cache.put(key.to_string(), val.clone());
        Some(val)
    } else {
        None
    }
}

fn main() {
    let mut cache = Cache::new(Duration::from_millis(300));
    let db = Database::new();

    // First access — cache miss
    println!("=== First access ===");
    let user = get_user(&mut cache, &db, "user:1");
    println!("Result: {:?}\n", user);

    // Second access — cache hit
    println!("=== Second access ===");
    let user = get_user(&mut cache, &db, "user:1");
    println!("Result: {:?}\n", user);

    // Invalidate on write
    println!("=== After invalidation ===");
    cache.invalidate("user:1");
    let user = get_user(&mut cache, &db, "user:1");
    println!("Result: {:?}\n", user);

    // TTL expiration
    println!("=== After TTL expiry ===");
    std::thread::sleep(Duration::from_millis(350));
    let user = get_user(&mut cache, &db, "user:1");
    println!("Result: {:?}", user);
}
